### ISSUE TITLE
jenkins integration issue - how to deploy application after build

### ISSUE THREAD

**Issue by tenstriker** [play2-maven-plugin/play2-maven-plugin#161]
How do you deploy play application after jenkins build? I added Post build step for Execute shell with following commands

```
#!/bin/sh
cd jenkins/workspace/myproject
nohup mvn play2:run -X -Dhttp.port=1234 & 
```

It's not starting the server. I don't see any error in jenkins log either.

...
Finished Calculation of disk usage of build in 0 seconds
Started calculate disk usage of workspace
Finished Calculation of disk usage of workspace in 0 seconds
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 10.874 s
[INFO] Finished at: 2018-01-26T15:53:14-08:00
[INFO] Final Memory: 29M/424M
[INFO] ------------------------------------------------------------------------
[myproject] $ /bin/sh /tmp/hudson3154357614303774301.sh
Publishing Scoverage XML and HTML report ...
Started calculate disk usage of build
Finished Calculation of disk usage of build in 0 seconds
Started calculate disk usage of workspace
Finished Calculation of disk usage of workspace in 0 seconds
Finished: SUCCESS

**Comment by gslowikowski** [comment]
Hi @tenstriker 

Where do you want to deploy it? Additionally, what exactly do you want to deploy, is important question too.

If you want to prepare and deploy a complete distribution, ready to be run (with all dependencies), use `play2:dist` goal or create custom distribution with Maven Assembly Plugin like in [this example project](https://github.com/gslowikowski/play2-scala-dist-using-assembly) .

Don't use `play:run` goal in CI environment for some reasons:
- it's for development purposes, allows running the application while still changing the code
- it's interactive (exits after pressing [Enter]), so your build will never end (or will timeout) because there is nobody to hit [Enter]

Additionally, it doesn't create any kind of distribution.

**Comment by tenstriker** [comment]
I am trying to deploy inside same Jenkins slave machine where the build happened. Maven is instlled there as well. Assembly shouldn't be required there. `mvn play2:start` seems viable solution here but issue is it's not taking in command line java param. I need to change port when starting server. I tried ` mvn play2:start -Dhttp.port=1234` but it is still trying to start it on 9000 and failing as I have HA proxy running on 9000. How do I specify cli options with `play2: start`

**Comment by gslowikowski** [comment]
I don't know, if Jenkins slave machine is the best choice to start your application, but if you want, use [prodSettings](https://play2-maven-plugin.github.io/play2-maven-plugin/1.0.0-rc2/play2-maven-plugin/start-mojo.html#prodSettings) parameter:
```
nohup mvn play2:start -DprodSettings=play.server.http.port=1234 &
```

`run` goal has [devSettings](https://play2-maven-plugin.github.io/play2-maven-plugin/1.0.0-rc2/play2-maven-plugin/run-mojo.html#devSettings) parameter, like [in Play SBT plugin](https://www.playframework.com/documentation/2.6.x/ConfigFile#Extra-devSettings)
`start` goal has [prodSettings](https://play2-maven-plugin.github.io/play2-maven-plugin/1.0.0-rc2/play2-maven-plugin/start-mojo.html#prodSettings) parameter, no similar parameter in Play SBT plugin

**Comment by tenstriker** [comment]
I tried this exact command:
`nohup mvn play2:start -DprodSettings=play.server.http.port=1234 &`

It just keep starting on 9000 and completely ignores command line parameter. 

Can you verify it works on 1.0.0-rc2 version?

**Comment by gslowikowski** [comment]
The parameter setting the port should be `-Dplay2.prodSettings=play.server.http.port=1234`.

When I've tested this on a starter project, application didn't start at all and there are no logs. This is because the forked process must be disconnected from the Maven process to survive it's end.
So, application doesn't start, and I cannot check, why.

The solution is to start it without spawning. You have to add some more magic parameters (I must simplify it somehow):

`mvn play2:start -Dplay2.prodSettings=play.server.http.port=1234 -Dplay2.startSpawn=false -Dplay2.startSynchro -Dplay2.startCheckUrl=http://localhost:1234/`

I my case I saw the error like this:
```
[INFO] Starting Play! server
[INFO] Waiting for http://localhost:1234/
     [play] Oops, cannot start the server.
     [play] Configuration error: Configuration error[
     [play] The application secret has not been set, and we are in prod mode. Your application is not secure.
     [play] To set the application secret, please read http://playframework.com/documentation/latest/ApplicationSecret
     [play]           ]
     [play]     at play.api.Configuration$.configError(Configuration.scala:156)
...
```

It means, that I have to add, or change the default `changeme` value to something else, of the `play.http.secret.key` (in Play 2.6.x, in previous versions key was different).

After I changed the secret value, rebuilt the app (for the changed `application.conf` file to be copied to `target/classes` directory) and restart the application is started without problems and stopped immediately, when Maven build process stopped.

This is OK. Now the extra parameters `-Dplay2.startSpawn=false -Dplay2.startSynchro -Dplay2.startCheckUrl=http://localhost:1234/` must be removed and application should start successfully.

All of this is too complicated, I know. I will think, how to make it easier.

**Comment by tenstriker** [comment]
I am little confuse now how `The application secret has not been set` error related to play2:start not accepting command line parameter.
Spawing issue that your mention sound related but the error `The application secret has not been set` is unrelated to command line option problem. I had that error as well initially in prod mode even with default port. I did get rid of it by adding that secret key. 
Can you summarize what is the workaround now? Do I need to use `-Dplay2.startSpawn=false -Dplay2.startSynchro -Dplay2.startCheckUrl=http://localhost:1234/`  or not?
I tried it and seems like it is starting on 1234 with above option but only issue is it constantly checks http://localhost:1234/ url where I have nothing so I keep getting 404. I removed `-Dplay2.startCheckUrl` and now it's working . I just have to put process in background since it's not spawnig anymore.

**Comment by gslowikowski** [comment]
The tricks with turning off spawning are only to see on the console the error preventing the application start. When I fixed the problem (with application secret) I returned to the initial command
```
mvn play2:start -Dplay2.prodSettings=play.server.http.port=1234
```
and it worked. I was able to access my application at `http://localhost:1234`.

Can you check it on lour local machine before testing in Jenkins?

**Comment by tenstriker** [comment]
it does work now for some reason! I think property name has to be `play2.prodSettings=play.server.http.port=1234` and not `-DprodSettings=play.server.http.port=1234`
earlier we were using  just `prodSettings` without play2 prefix. 
Thanks for spawning trick. Could be handy when server doesn't start.
